### PR TITLE
Allow validators to register address to send report transactions - Declare messages

### LINKED ISSUE #847: Allow validators to register address to send report transactions - Declare messages
Add two messages: `MsgSetOracleAddress` and `MsgUnsetOracleAddress`. The messages allow validators to set the address that can be the signer of its report transaction.

### PR DESCRIPTION
**Pull request by Benzbeeb**
Fixed #847

### CODE REVIEW

File: chain/x/zoracle/internal/types/msgs.go [hunk 1]
```diff
@@ -428,3 +428,95 @@ func (msg MsgEditOracleScript) GetSignBytes() []byte {
 	bz := ModuleCdc.MustMarshalJSON(msg)
 	return sdk.MustSortJSON(bz)
 }
+
+// MsgSetOracleAddress is a message for set a agent to report transactions.
+type MsgSetOracleAddress struct {
```
Comment (sorawit) [inline chain/x/zoracle/internal/types/msgs.go]: Let's call this `MsgAddOracleAddress` and `MsgRemoveOracleAdderess`

File: chain/x/zoracle/internal/types/msgs.go [hunk 2]
```diff
@@ -428,3 +428,95 @@ func (msg MsgEditOracleScript) GetSignBytes() []byte {
 	bz := ModuleCdc.MustMarshalJSON(msg)
 	return sdk.MustSortJSON(bz)
 }
+
+// MsgSetOracleAddress is a message for set a agent to report transactions.
```
Comment (sorawit) [inline chain/x/zoracle/internal/types/msgs.go]: for adding an agent authorized to submit report transactions

File: chain/x/zoracle/internal/types/msgs.go [hunk 3]
```diff
@@ -428,3 +428,95 @@ func (msg MsgEditOracleScript) GetSignBytes() []byte {
...
+}
+
+// GetSignBytes implements the sdk.Msg interface for MsgSetOracleAddress.
+func (msg MsgSetOracleAddress) GetSignBytes() []byte {
+	bz := ModuleCdc.MustMarshalJSON(msg)
+	return sdk.MustSortJSON(bz)
+}
+
+// MsgUnsetOracleAddress is a message for set a agent to report transactions.
+type MsgUnsetOracleAddress struct {
```
Comment (sorawit) [inline chain/x/zoracle/internal/types/msgs.go]: for removing an agent from the list of authorized reporters

File: chain/x/zoracle/internal/types/msgs.go [hunk 4]
```diff
@@ -428,3 +428,95 @@ func (msg MsgEditOracleScript) GetSignBytes() []byte {
...
+		return ErrInvalidBasicMsg("MsgSetOracleAddress: Validator address must not be empty.")
+	}
+	if msg.Reporter.Empty() {
+		return ErrInvalidBasicMsg("MsgSetOracleAddress: Reporter address must not be empty.")
+	}
+	return nil
+}
+
+// GetSigners implements the sdk.Msg interface for MsgSetOracleAddress.
+func (msg MsgSetOracleAddress) GetSigners() []sdk.ValAddress {
```
Comment (sorawit) [inline chain/x/zoracle/internal/types/msgs.go]: Should return `[]sdk.AccAddress`. Also ditto below.

File: chain/x/zoracle/internal/types/msgs.go [hunk 5]
```diff
@@ -428,3 +428,95 @@ func (msg MsgEditOracleScript) GetSignBytes() []byte {
...
+		Validator: validator,
+		Reporter:  reporter,
+	}
+}
+
+// Route implements the sdk.Msg interface for MsgSetOracleAddress.
+func (msg MsgSetOracleAddress) Route() string { return RouterKey }
+
+// Type implements the sdk.Msg interface for MsgSetOracleAddress.
+func (msg MsgSetOracleAddress) Type() string { return "set_oracle_script_address" }
```
Comment (sorawit) [inline chain/x/zoracle/internal/types/msgs.go]: No "script"

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/zoracle/internal/types/msgs.go [hunk 6]
```diff
@@ -428,3 +428,95 @@ func (msg MsgEditOracleScript) GetSignBytes() []byte {
...
+		Validator: validator,
+		Reporter:  reporter,
+	}
+}
+
+// Route implements the sdk.Msg interface for MsgUnsetOracleAddress.
+func (msg MsgUnsetOracleAddress) Route() string { return RouterKey }
+
+// Type implements the sdk.Msg interface for MsgUnsetOracleAddress.
+func (msg MsgUnsetOracleAddress) Type() string { return "unset_oracle_script_address" }
```
Comment (taobun) [inline chain/x/zoracle/internal/types/msgs.go]: ditto

File: chain/x/zoracle/internal/types/msgs_test.go [hunk 7]
```diff
@@ -480,3 +480,23 @@ func TestMsgEditOracleScriptGetSignBytes(t *testing.T) {
...
+func TestMsgSetOracleAddress(t *testing.T) {
+	validator := sdk.ValAddress([]byte("validator"))
+	reporter := sdk.AccAddress([]byte("reporter"))
+	msg := NewMsgSetOracleAddress(validator, reporter)
+	require.Equal(t, RouterKey, msg.Route())
+	require.Equal(t, "set_oracle_script_address", msg.Type())
+	require.Equal(t, validator, msg.Validator)
+	require.Equal(t, reporter, msg.Reporter)
+}
+
```
Comment (taobun) [inline chain/x/zoracle/internal/types/msgs_test.go:488]: Test validate basic and getSignBytes too

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/x/zoracle/internal/types/msgs_test.go (hunk 7 above)
Comment (sorawit) [reply chain/x/zoracle/internal/types/msgs_test.go:488]: @Benzbeeb Why is this not done yet?

**Review by sorawit — CHANGES_REQUESTED** [review]

**Review by sorawit — DISMISSED** [review]
LGTM

File: chain/x/zoracle/internal/types/msgs_test.go [hunk 8]
```diff
@@ -480,3 +480,83 @@ func TestMsgEditOracleScriptGetSignBytes(t *testing.T) {
...
+func TestMsgSetOracleAddress(t *testing.T) {
+	validator := sdk.ValAddress([]byte("validator"))
+	reporter := sdk.AccAddress([]byte("reporter"))
+	msg := NewMsgAddOracleAddress(validator, reporter)
+	require.Equal(t, RouterKey, msg.Route())
+	require.Equal(t, "add_oracle_address", msg.Type())
+	require.Equal(t, validator, msg.Validator)
+	require.Equal(t, reporter, msg.Reporter)
+
+	cases := []struct {
```
Comment (sorawit) [inline chain/x/zoracle/internal/types/msgs_test.go:493]: Normally we separate this into different unit tests? Opinion @taobun ?

Comment (taobun) [reply chain/x/zoracle/internal/types/msgs_test.go:493]: Let's break it

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/x/zoracle/internal/types/msgs_test.go [hunk 9]
```diff
@@ -480,3 +480,93 @@ func TestMsgEditOracleScriptGetSignBytes(t *testing.T) {
...
+	validator := sdk.ValAddress([]byte("validator"))
+	reporter := sdk.AccAddress([]byte("reporter"))
+	msg := NewMsgRemoveOracleAdderess(validator, reporter)
+	require.Equal(t, RouterKey, msg.Route())
+	require.Equal(t, "remove_oracle_address", msg.Type())
+	require.Equal(t, validator, msg.Validator)
+	require.Equal(t, reporter, msg.Reporter)
+	res := msg.GetSignBytes()
+
+	expected := `{"reporter":"band1wfjhqmmjw3jhyy3as6w","validator":"bandvaloper1weskc6tyv96x7usd82k92"}`
```
Comment (taobun) [inline chain/x/zoracle/internal/types/msgs_test.go]: Move this test to another unit test

**Review by taobun — CHANGES_REQUESTED** [review]

**Review by taobun — APPROVED** [review]

**Review by sorawit — APPROVED** [review]

**Pull request merged**